hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestCOWDataSource.scala

```diff
@@ -1410,6 +1411,47 @@ class TestCOWDataSource extends HoodieSparkClientTestBase with ScalaAssertionSup
     assertEquals(false, Metrics.isInitialized(basePath), "Metrics should be shutdown")
   }
 
+  /**
+   * Validates that clustering dag is triggered only once.
+   * We leverage spark event listener to validate it.
+   */
+  @Test
+  def testValidateClusteringForRepeatedDag(): Unit = {
```

Comment:
Note to Reviewer: this test fails w/o the fix in this patch. w/ the fix, it succeeds.